sklearn/utils/validation.py

```diff
@@ -232,25 +234,27 @@ def _ensure_sparse_format(spmatrix, accept_sparse, dtype, copy,
     spmatrix_converted : scipy sparse matrix.
         Matrix that is ensured to have an allowed type.
     """
-    if accept_sparse is None:
+    if accept_sparse in [None, False]:
         raise TypeError('A sparse matrix was passed, but dense '
                         'data is required. Use X.toarray() to '
                         'convert to a dense numpy array.')
-    sparse_type = spmatrix.format
     if dtype is None:
         dtype = spmatrix.dtype
-    if sparse_type in accept_sparse:
-        # correct type
-        if dtype == spmatrix.dtype:
-            # correct dtype
-            if copy:
-                spmatrix = spmatrix.copy()
-        else:
-            # convert dtype
-            spmatrix = spmatrix.astype(dtype)
-    else:
-        # create new
-        spmatrix = spmatrix.asformat(accept_sparse[0]).astype(dtype)
+
+    changed_format = False
+    if (isinstance(accept_sparse, (list, tuple))
+            and spmatrix.format not in accept_sparse):
+        # create new with correct sparse
+        spmatrix = spmatrix.asformat(accept_sparse[0])
+        changed_format = True
+
+    if dtype != spmatrix.dtype:
+        # convert dtype
+        spmatrix = spmatrix.astype(dtype)
+    elif copy and not changed_format:
```

Comment:
or maybe not actually, because when we ask for a copy we might expect the other parts of the scipy sparse dastructure to be copied as well, not just the array that store the dtype.